parser/parser_resources.go

```diff
@@ -111,63 +29,64 @@ type resourceCreationVisitor struct {
-								walkSpec = false
+						ident = spec.Names[0]
+					} else {
+						f.p.errf(
+							spec.Pos(),
+							"A %s must be bound to a variable with one name and value. For more information please see %s",
+							parser.Resource.Name,
+							parser.Resource.Docs,
+						)
+						return nil
+					}
+				}
 
-								parser.Parse(f.p, f.file, docNodeToString(node.Doc), spec, callExpr)
-							}
+				// If the parser allows resource to be created here, let's call parse it
+				// and then record the resource that was created
+				if resource := parser.Parse(f.p, f.file, cursor, ident, node); resource != nil {
+
+					if ident != nil {
+						f.file.References[ident] = &est.Node{
+							Type: resource.NodeType(),
+							Res:  resource,
 						}
 					}
-				}
 
-				if walkSpec {
-					ast.Walk(walkerFunc(f.VisitAndReportInvalidCreationCalls), spec)
+					f.file.Pkg.Resources = append(f.file.Pkg.Resources, resource)
 				}
+			} else {
+				f.p.errf(
+					node.Pos(),
+					"A %s can not be declared here, %s. For more information please see %s",
+					parser.Resource.Name,
+					parser.AllowedLocations.Describe("they", "declared"),
+					parser.Resource.Docs,
+				)
 			}
 
-			// We don't want to visit the GenDecl node as we've already manually walked it, so we return nil here.
 			return nil
 		}
 
-		return walkerFunc(f.VisitAndReportInvalidCreationCalls)
-	case *ast.FuncDecl:
-		return walkerFunc(f.VisitAndReportInvalidCreationCalls)
-	default:
-		return f
-	}
-}
-
-// VisitAndReportInvalidCreationCalls walks the AST looking for calls to resource creation function, however
-// if we encounter them, then we need to report an error as the resource creation function is only allowed to be used
-// as a top-level variable declaration.
-func (f *resourceCreationVisitor) VisitAndReportInvalidCreationCalls(node ast.Node) bool {
-	switch node := node.(type) {
-	case *ast.CallExpr:
-		if parser := f.parserFor(node.Fun); parser != nil {
-			f.p.errf(node.Pos(), "A %s must be declared as a package level variable. For more information please see %s\n", parser.Resource.Name, parser.Resource.Docs)
-			return false
-		}
-
 	case *ast.SelectorExpr, *ast.IndexExpr, *ast.IndexListExpr:
+		// If we find a selector (`a.foo`) an index (`a.foo[bar]`) or an index list (`a.foo[bar, baz]`)
+		// then we want to check if that references a resource creation function and if so
+		// report an error, as all valid usages should already have been parsed and returned
 		if parser := f.parserFor(node); parser != nil {
-			f.p.errf(node.Pos(), "%s.%s can only be used to declare package level variables. For more information please see %s\n", parser.Resource.PkgName, parser.Name, parser.Resource.Docs)
-			return false
+			f.p.errf(node.Pos(), "%s.%s can only be called as a function to create a new instance and not referenced otherwise. For more information please see %s", parser.Resource.PkgName, parser.Name, parser.Resource.Docs)
```

Comment:
```suggestion
			f.p.errf(node.Pos(), "%s.%s can only be called as a function to create a new instance and not referenced otherwise. For more information see %s", parser.Resource.PkgName, parser.Name, parser.Resource.Docs)
```### ###